File: sort.java
import java.util.Arrays;
import java.io.*;
import java.util.Scanner;

public class sort {
    public static void insertionSort(int[] array) {
        for (int i = 1; i < array.length; i++) {
            int current = array[i];
            int j = i - 1;
            while (j >= 0 && current < array[j]) {
                array[j + 1] = array[j];
                j--;
            }
            array[j + 1] = current;
        }
    }

    public void mergeSort(int[] array, int left, int right) {
        if (right <= left) return;
        int mid = (left+right)/2;
        mergeSort(array, left, mid);
        mergeSort(array, mid+1, right);
        merge(array, left, mid, right);
    }

    void merge(int[] array, int left, int mid, int right) {
        int lengthLeft = mid - left + 1;
        int lengthRight = right - mid;

        int leftArray[] = new int [lengthLeft];
        int rightArray[] = new int [lengthRight];

        for (int i = 0; i < lengthLeft; i++)
            leftArray[i] = array[left+i];
        for (int i = 0; i < lengthRight; i++)
            rightArray[i] = array[mid+i+1];

        int leftIndex = 0;
        int rightIndex = 0;

        for (int i = left; i < right + 1; i++) {
            if (leftIndex < lengthLeft && rightIndex < lengthRight) {
                if (leftArray[leftIndex] < rightArray[rightIndex]) {
                    array[i] = leftArray[leftIndex];
                    leftIndex++;
                }
                else {
                    array[i] = rightArray[rightIndex];
                    rightIndex++;
                }
            }
            else if (leftIndex < lengthLeft) {
                array[i] = leftArray[leftIndex];
                leftIndex++;
            }
            else if (rightIndex < lengthRight) {
                array[i] = rightArray[rightIndex];
                rightIndex++;
            }
        }
    }

    int[] countElements(int[] input, int k) {
        int[] c = new int[k + 1];
        Arrays.fill(c, 0);
        for (int i : input) {
            c[i] += 1;
        }

        for (int i = 1; i < c.length; i++) {
            c[i] += c[i - 1];
        }

        return c;
    }

    public void countingSort(int[] arr, int range) {
        int[] count = new int[range];
        Arrays.fill(count, 0);
        for (int i : arr) {
            count[i] += 1;
        }

        int k = 0;
        for (int i = 0; i < range; i++){
            while(count[i] > 0){
                arr[k] = i;
                k++;
                count[i]--;
            }
        }
    }

    public static void main(String[] args) throws IOException {

        final int N = 1000;
        File array = new File("array.txt");
        Scanner scanner = new Scanner(array);
        int [] arr = new int [N];
        int i = 0;
        long start, end;
        double insertion_time, merge_time, counting_time;

        while(scanner.hasNextInt()){
            arr[i++] = scanner.nextInt();
        }



        start = System.nanoTime();
       insertionSort(arr);
        end = System.nanoTime();
        insertion_time = (double) (end - start) / 1_000_000_000;

        i = 0;
        scanner = new Scanner(array);
        while(scanner.hasNextInt()){
            arr[i++] = scanner.nextInt();
        }

        sort obj = new sort();
        start = System.nanoTime();
       obj.mergeSort(arr, 0, arr.length - 1);
        end = System.nanoTime();

        merge_time = (double) (end - start) / 1_000_000_000;

        i = 0;
        scanner = new Scanner(array);
        while(scanner.hasNextInt()){
            arr[i++] = scanner.nextInt();
        }

        start = System.nanoTime();
        obj.countingSort(arr, 1000);
        end = System.nanoTime();
        counting_time = (double) (end - start) / 1_000_000_000;


        // System.out.println(Arrays.toString(arr));
        // System.out.println("Hello, I am java file!");

       String output = String.format("Java: Insertion sort time: %f\n" +
               "Java: Merge sort time: %f\n" +
               "Java: Counting sort time: %f\n", insertion_time, merge_time, counting_time);
       File file = new File("time-measured.txt");
       FileWriter fr = new FileWriter(file, true);
       fr.write(output);
       fr.close();

    }

}

File: sort.kt
import java.util.*
import java.io.File

fun insertionSort(arr: IntArray): Unit{
    for (i in 1 until arr.size){
        val current = arr[i]
        var j = i - 1
        while(j >= 0 && current < arr[j]){
            arr[j + 1] = arr[j]
            j--
        }
        arr[j + 1] = current
    }
}

fun merge(arr: IntArray, l: Int, m: Int, r: Int): Unit{
    val n1 = m - l + 1
    val n2 = r - m
    val L = IntArray(n1)
    val R = IntArray(n2)
    for (i in 0 until n1){
        L[i] = arr[i + l]
    }
    for (j in 0 until n2){
        R[j] = arr[j + m + 1];
    }
    var i = 0
    var j = 0
    var k = l
    while (i < n1 && j < n2){
        if (L[i] <= R[j]){
            arr[k] = L[i]
            i++
        }
        else{
            arr[k] = R[j]
            j++
        }
        k++
    }
    while(i < n1){
        arr[k] = L[i]
        i++
        k++
    }
    while(j < n2){
        arr[k] = R[j]
        j++
        k++
    }
}


fun mergeSort(arr: IntArray, l: Int, r: Int){
    if (l < r){
        var m: Int = l + (r - l)/2
        mergeSort(arr, l, m)
        mergeSort(arr, m + 1, r)
        merge(arr, l, m, r)
    }
}




fun countingSort(arr: IntArray, range: Int = 1000): Unit{
    val count = IntArray(range)
    for (element in arr){
        count[element]++
    }
    var k = 0
    for (i in 0 until range){
        while (count[i] > 0){
            arr[k] = i
            count[i]--
            k++
        }
    }


}

fun main(args : Array<String>){

    val N = 1000
    val arrayContent = File("array.txt")
    var scanner = Scanner(arrayContent)
    val arr = IntArray(N)
    var i = 0
    var start: Long
    var end: Long
    var insertion_time: Double
    var merge_time: Double
    var counting_time: Double

    while (scanner.hasNextInt()) {
        arr[i++] = scanner.nextInt()
    }

    start = System.nanoTime()
    insertionSort(arr)
    end = System.nanoTime()
    insertion_time =  (end - start).toDouble() / 1_000_000_000

//    i = 0;
//    scanner = Scanner(arrayContent);
//    while(scanner.hasNextInt()){
//        arr[i++] = scanner.nextInt();
//    }

    start = System.nanoTime()
    mergeSort(arr, 0, arr.lastIndex)
    end = System.nanoTime()
    merge_time = (end - start).toDouble() / 1_000_000_000

    start = System.nanoTime()
    countingSort(arr)
    end = System.nanoTime()
    counting_time = (end - start).toDouble() / 1_000_000_000

    println("Insertion sort time: %f\n".format(insertion_time) +
            "Merge sort time: %f\n".format(merge_time) +
            "Counting sort time: %f\n".format(counting_time))
    // println("Hello, I am Kotlin file!\n");
}
